### PR TITLE
Add timeout flag to gRPC cli

### PR DESCRIPTION
**Pull request by ananda1066**
@karthikravis @pszemus

### CODE REVIEW

**Comment by karthikravis** [comment]
Please sign the CL @ananda1066

**Comment by ananda1066** [comment]
@karthikravis done 

@sheenaqotj, @pszemus  please review

**Comment by pszemus** [comment]
Cannot build using https://github.com/grpc/grpc/blob/master/doc/command_line_tool.md
```
[100%] Linking CXX executable grpc_cli
/usr/bin/ld: CMakeFiles/grpc_cli.dir/test/cpp/util/cli_call.cc.o: in function `grpc::testing::CliCall::CliCall(std::shared_ptr<grpc::Channel> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::multimap<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&)':
cli_call.cc:(.text+0x2d3): undefined reference to `grpc::testing::fLD::FLAGS_timeout'
/usr/bin/ld: cli_call.cc:(.text+0x2e5): undefined reference to `grpc::testing::fLD::FLAGS_timeout'
/usr/bin/ld: cli_call.cc:(.text+0x316): undefined reference to `grpc_timeout_microseconds_to_deadline(long)'
/usr/bin/ld: CMakeFiles/grpc_cli.dir/test/cpp/util/grpc_tool.cc.o: in function `grpc::testing::(anonymous namespace)::ParseMetadataFlag(std::multimap<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >*)':
grpc_tool.cc:(.text+0xaa): undefined reference to `grpc::testing::fLS::FLAGS_metadata[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0xf4): undefined reference to `grpc::testing::fLS::FLAGS_metadata[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x112): undefined reference to `grpc::testing::fLS::FLAGS_metadata[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x140): undefined reference to `grpc::testing::fLS::FLAGS_metadata[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x15e): undefined reference to `grpc::testing::fLS::FLAGS_metadata[abi:cxx11]'
/usr/bin/ld: CMakeFiles/grpc_cli.dir/test/cpp/util/grpc_tool.cc.o:grpc_tool.cc:(.text+0x291): more undefined references to `grpc::testing::fLS::FLAGS_metadata[abi:cxx11]' follow
/usr/bin/ld: CMakeFiles/grpc_cli.dir/test/cpp/util/grpc_tool.cc.o: in function `grpc::testing::(anonymous namespace)::ReadResponse(grpc::testing::CliCall*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::function<bool (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>, grpc::testing::ProtoFileParser*, long*, bool)':
grpc_tool.cc:(.text+0x54d): undefined reference to `grpc::testing::fLB::FLAGS_binary_output'
/usr/bin/ld: grpc_tool.cc:(.text+0x56e): undefined reference to `grpc::testing::fLB::FLAGS_json_output'
/usr/bin/ld: CMakeFiles/grpc_cli.dir/test/cpp/util/grpc_tool.cc.o: in function `grpc::testing::(anonymous namespace)::CreateCliChannel(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, grpc::testing::CliCredentials const&)':
grpc_tool.cc:(.text+0x80e): undefined reference to `grpc::testing::fLS::FLAGS_default_service_config[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x834): undefined reference to `grpc::testing::fLS::FLAGS_default_service_config[abi:cxx11]'
/usr/bin/ld: CMakeFiles/grpc_cli.dir/test/cpp/util/grpc_tool.cc.o: in function `grpc::testing::(anonymous namespace)::GrpcTool::ListServices(int, char const**, grpc::testing::CliCredentials const&, std::function<bool (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>)':
grpc_tool.cc:(.text+0x14d9): undefined reference to `grpc::testing::fLB::FLAGS_l'
/usr/bin/ld: grpc_tool.cc:(.text+0x179c): undefined reference to `grpc::testing::fLB::FLAGS_l'
/usr/bin/ld: grpc_tool.cc:(.text+0x185f): undefined reference to `grpc::testing::fLB::FLAGS_l'
/usr/bin/ld: grpc_tool.cc:(.text+0x1982): undefined reference to `grpc::testing::fLB::FLAGS_l'
/usr/bin/ld: CMakeFiles/grpc_cli.dir/test/cpp/util/grpc_tool.cc.o: in function `grpc::testing::(anonymous namespace)::GrpcTool::CallMethod(int, char const**, grpc::testing::CliCredentials const&, std::function<bool (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>)':
grpc_tool.cc:(.text+0x2134): undefined reference to `grpc::testing::fLB::FLAGS_binary_input'
/usr/bin/ld: grpc_tool.cc:(.text+0x2142): undefined reference to `grpc::testing::fLB::FLAGS_binary_output'
/usr/bin/ld: grpc_tool.cc:(.text+0x2154): undefined reference to `grpc::testing::fLB::FLAGS_remotedb'
/usr/bin/ld: grpc_tool.cc:(.text+0x21a2): undefined reference to `grpc::testing::fLS::FLAGS_protofiles[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x21a9): undefined reference to `grpc::testing::fLS::FLAGS_proto_path[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x222f): undefined reference to `grpc::testing::fLB::FLAGS_binary_input'
/usr/bin/ld: grpc_tool.cc:(.text+0x235c): undefined reference to `grpc::testing::fLB::FLAGS_batch'
/usr/bin/ld: grpc_tool.cc:(.text+0x2436): undefined reference to `grpc::testing::fLB::FLAGS_display_peer_address'
/usr/bin/ld: grpc_tool.cc:(.text+0x24ab): undefined reference to `grpc::testing::fLS::FLAGS_infile[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x2522): undefined reference to `grpc::testing::fLS::FLAGS_infile[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x268b): undefined reference to `grpc::testing::fLB::FLAGS_binary_input'
/usr/bin/ld: grpc_tool.cc:(.text+0x26e4): undefined reference to `grpc::testing::fLB::FLAGS_json_input'
/usr/bin/ld: grpc_tool.cc:(.text+0x2ae5): undefined reference to `grpc::testing::fLB::FLAGS_batch'
/usr/bin/ld: grpc_tool.cc:(.text+0x2b5a): undefined reference to `grpc::testing::fLS::FLAGS_infile[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x2bd1): undefined reference to `grpc::testing::fLS::FLAGS_infile[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x2d39): undefined reference to `grpc::testing::fLB::FLAGS_binary_input'
/usr/bin/ld: grpc_tool.cc:(.text+0x2d83): undefined reference to `grpc::testing::fLB::FLAGS_json_input'
/usr/bin/ld: grpc_tool.cc:(.text+0x2e8a): undefined reference to `grpc::testing::fLB::FLAGS_display_peer_address'
/usr/bin/ld: grpc_tool.cc:(.text+0x309a): undefined reference to `grpc::testing::fLB::FLAGS_binary_output'
/usr/bin/ld: grpc_tool.cc:(.text+0x30e5): undefined reference to `grpc::testing::fLB::FLAGS_json_output'
/usr/bin/ld: grpc_tool.cc:(.text+0x341f): undefined reference to `grpc::testing::fLS::FLAGS_infile[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x346d): undefined reference to `grpc::testing::fLS::FLAGS_infile[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x34f7): undefined reference to `grpc::testing::fLS::FLAGS_infile[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x35a6): undefined reference to `grpc::testing::fLB::FLAGS_binary_input'
/usr/bin/ld: grpc_tool.cc:(.text+0x35e1): undefined reference to `grpc::testing::fLB::FLAGS_json_input'
/usr/bin/ld: grpc_tool.cc:(.text+0x377d): undefined reference to `grpc::testing::fLB::FLAGS_display_peer_address'
/usr/bin/ld: grpc_tool.cc:(.text+0x3853): undefined reference to `grpc::testing::fLB::FLAGS_binary_output'
/usr/bin/ld: grpc_tool.cc:(.text+0x3877): undefined reference to `grpc::testing::fLB::FLAGS_json_output'
/usr/bin/ld: CMakeFiles/grpc_cli.dir/test/cpp/util/grpc_tool.cc.o: in function `grpc::testing::(anonymous namespace)::GrpcTool::ParseMessage(int, char const**, grpc::testing::CliCredentials const&, std::function<bool (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>)':
grpc_tool.cc:(.text+0x4217): undefined reference to `grpc::testing::fLS::FLAGS_infile[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x4265): undefined reference to `grpc::testing::fLS::FLAGS_infile[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x42ef): undefined reference to `grpc::testing::fLS::FLAGS_infile[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x439e): undefined reference to `grpc::testing::fLB::FLAGS_binary_input'
/usr/bin/ld: grpc_tool.cc:(.text+0x43ac): undefined reference to `grpc::testing::fLB::FLAGS_binary_output'
/usr/bin/ld: grpc_tool.cc:(.text+0x43de): undefined reference to `grpc::testing::fLB::FLAGS_remotedb'
/usr/bin/ld: grpc_tool.cc:(.text+0x4423): undefined reference to `grpc::testing::fLS::FLAGS_protofiles[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x442a): undefined reference to `grpc::testing::fLS::FLAGS_proto_path[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x44cf): undefined reference to `grpc::testing::fLB::FLAGS_binary_input'
/usr/bin/ld: grpc_tool.cc:(.text+0x450a): undefined reference to `grpc::testing::fLB::FLAGS_json_input'
/usr/bin/ld: grpc_tool.cc:(.text+0x4598): undefined reference to `grpc::testing::fLB::FLAGS_binary_output'
/usr/bin/ld: grpc_tool.cc:(.text+0x45e6): undefined reference to `grpc::testing::fLB::FLAGS_json_output'
/usr/bin/ld: CMakeFiles/grpc_cli.dir/test/cpp/util/grpc_tool.cc.o: in function `grpc::testing::(anonymous namespace)::GrpcTool::ToText(int, char const**, grpc::testing::CliCredentials const&, std::function<bool (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>)':
grpc_tool.cc:(.text+0x492e): undefined reference to `grpc::testing::fLS::FLAGS_protofiles[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x493f): undefined reference to `grpc::testing::fLB::FLAGS_remotedb'
/usr/bin/ld: grpc_tool.cc:(.text+0x4946): undefined reference to `grpc::testing::fLB::FLAGS_binary_input'
/usr/bin/ld: grpc_tool.cc:(.text+0x494d): undefined reference to `grpc::testing::fLB::FLAGS_binary_output'
/usr/bin/ld: CMakeFiles/grpc_cli.dir/test/cpp/util/grpc_tool.cc.o: in function `grpc::testing::(anonymous namespace)::GrpcTool::ToJson(int, char const**, grpc::testing::CliCredentials const&, std::function<bool (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>)':
grpc_tool.cc:(.text+0x4a62): undefined reference to `grpc::testing::fLS::FLAGS_protofiles[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x4a73): undefined reference to `grpc::testing::fLB::FLAGS_remotedb'
/usr/bin/ld: grpc_tool.cc:(.text+0x4a7a): undefined reference to `grpc::testing::fLB::FLAGS_binary_input'
/usr/bin/ld: grpc_tool.cc:(.text+0x4a81): undefined reference to `grpc::testing::fLB::FLAGS_binary_output'
/usr/bin/ld: grpc_tool.cc:(.text+0x4a88): undefined reference to `grpc::testing::fLB::FLAGS_json_output'
/usr/bin/ld: CMakeFiles/grpc_cli.dir/test/cpp/util/grpc_tool.cc.o: in function `grpc::testing::(anonymous namespace)::GrpcTool::ToBinary(int, char const**, grpc::testing::CliCredentials const&, std::function<bool (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>)':
grpc_tool.cc:(.text+0x4b9e): undefined reference to `grpc::testing::fLS::FLAGS_protofiles[abi:cxx11]'
/usr/bin/ld: grpc_tool.cc:(.text+0x4baf): undefined reference to `grpc::testing::fLB::FLAGS_remotedb'
/usr/bin/ld: grpc_tool.cc:(.text+0x4bb6): undefined reference to `grpc::testing::fLB::FLAGS_binary_input'
/usr/bin/ld: grpc_tool.cc:(.text+0x4bbd): undefined reference to `grpc::testing::fLB::FLAGS_binary_output'
collect2: error: ld returned 1 exit status
make[3]: *** [CMakeFiles/grpc_cli.dir/build.make:316: grpc_cli] Error 1
make[2]: *** [CMakeFiles/Makefile2:5170: CMakeFiles/grpc_cli.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:5177: CMakeFiles/grpc_cli.dir/rule] Error 2
make: *** [Makefile:899: grpc_cli] Error 2
```

**Comment by ananda1066** [comment]
@pszemus should be fixed now

File: test/cpp/util/cli_call.cc [hunk 1]
```diff
@@ -61,6 +62,19 @@ CliCall::CliCall(const std::shared_ptr<grpc::Channel>& channel,
       ctx_.AddMetadata(iter->first, iter->second);
     }
   }
+
+  // Set deadline if timeout > 0 (default value -1 if no timeout specified)
+  if (FLAGS_timeout > 0) {
+    int64_t timeout_in_us = ceil(FLAGS_timeout * 1e6);
```
Comment (sheenaqotj) [inline test/cpp/util/cli_call.cc:68]: Seems like since you're multiplying by 1e3 again, maybe just multiple by 1e9 and use timeout_in_nanos?

**Review by sheenaqotj — APPROVED** [review]

File: test/cpp/util/cli_call.cc (hunk 1 above)
Comment (ananda1066) [reply test/cpp/util/cli_call.cc]: Done.

**Comment by pszemus** [comment]
@ananda1066 I've checked the `--timeout` option and it works well with my gRPC server

**Comment by ananda1066** [comment]
Test failure is an infrastructure timeout.

**Pull request merged**